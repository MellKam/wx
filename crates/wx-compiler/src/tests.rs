use super::*;
use indoc::indoc;
use insta::{assert_binary_snapshot, assert_yaml_snapshot};
use string_interner::{StringInterner, backend::StringBackend};

#[allow(unused)]
struct TestCase {
    files: files::Files,
    interner: StringInterner<StringBackend>,
    ast: ast::parser::ParserResult,
    hir: hir::builder::BuilderResult,
    mir: mir::MIR,
    module: wasm::Module,
    bytecode: Vec<u8>,
}

impl TestCase {
    fn assert_snapshot(name: &str, source: &str) {
        let mut interner = StringInterner::new();
        let mut files = files::Files::new();
        let file_id = files
            .add(name.to_string() + ".wx", source.to_string())
            .unwrap();

        let ast =
            ast::parser::Parser::parse(file_id, &files.get(file_id).unwrap().source, &mut interner);
        assert_yaml_snapshot!(name.to_string() + "_ast", ast);

        if ast.diagnostics.len() > 0 {
            return;
        }

        let hir = hir::Builder::build(&ast.ast, &mut interner);
        assert_yaml_snapshot!(name.to_string() + "_hir", hir);

        let mir = mir::Builder::build(&hir.hir);
        assert_yaml_snapshot!(name.to_string() + "_mir", mir);

        let module = wasm::Builder::build(&mir).unwrap();
        assert_yaml_snapshot!(name.to_string() + "_wasm", module);

        let bytecode = wasm::Encoder::encode(&module, &interner);
        assert_binary_snapshot!((name.to_string() + ".wasm").as_ref(), bytecode.clone());
    }
}

#[test]
fn factorial() {
    TestCase::assert_snapshot(
        "factorial",
        indoc! {"
        export func factorial(n: i32): i32 {
            if n <= 1 { return 1 };
            n * factorial(n - 1)
        }
    "},
    );
}

#[test]
fn pow() {
    TestCase::assert_snapshot(
        "pow",
        indoc! {"
        export func pow(base: i32, mut exp: i32): i32 {
            if exp < 0 {
                return 0; 
            };
            
            local mut result: i32 = 1;
            loop {
                if exp == 0 {
                break result; 
                };
                
                result *= base;
                exp -= 1;
            }
        }
    "},
    );
}

#[test]
fn func_pointers() {
    TestCase::assert_snapshot(
        "func_pointers",
        indoc! {"
        func add(a: i32, b: i32): i32 { a + b }
        func sub(a: i32, b: i32): i32 { a - b }

        func apply(
            binop: func(i32, i32) -> i32, 
            a: i32, 
            b: i32,
        ): i32 {
            binop(a, b)
        }

        export func main(): i32 {
            local a = apply(add, 5, 10);
            local b = apply(sub, 10, 5);
            local c: i32 = 5;

            a + b
        }
    "},
    );
}
