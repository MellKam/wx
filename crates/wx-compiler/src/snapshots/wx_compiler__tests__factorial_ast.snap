---
source: crates/wx-compiler/src/tests.rs
expression: ast
---
ast:
  file_id: 0
  items:
    - kind:
        ExportModifier:
          export:
            start: 0
            end: 6
          item:
            kind:
              FunctionDefinition:
                signature:
                  name:
                    symbol: 1
                    span:
                      start: 12
                      end: 21
                  params:
                    open:
                      start: 21
                      end: 22
                    inner:
                      - inner:
                          mutable: ~
                          name:
                            symbol: 2
                            span:
                              start: 22
                              end: 23
                          annotation:
                            separator:
                              start: 23
                              end: 24
                            ty:
                              kind:
                                Identifier:
                                  symbol: 3
                              span:
                                start: 25
                                end: 28
                        span:
                          start: 22
                          end: 28
                        separator: ~
                    close:
                      start: 28
                      end: 29
                  result:
                    separator:
                      start: 29
                      end: 30
                    ty:
                      kind:
                        Identifier:
                          symbol: 3
                      span:
                        start: 31
                        end: 34
                block:
                  kind:
                    Block:
                      open:
                        start: 35
                        end: 36
                      inner:
                        - inner:
                            Expression:
                              expr:
                                kind:
                                  IfElse:
                                    condition:
                                      kind:
                                        Binary:
                                          left:
                                            kind:
                                              Identifier:
                                                symbol: 2
                                            span:
                                              start: 44
                                              end: 45
                                          operator:
                                            kind: LessEq
                                            span:
                                              start: 46
                                              end: 48
                                          right:
                                            kind:
                                              Int:
                                                value: 1
                                            span:
                                              start: 49
                                              end: 50
                                      span:
                                        start: 44
                                        end: 50
                                    then_block:
                                      kind:
                                        Block:
                                          open:
                                            start: 51
                                            end: 52
                                          inner:
                                            - inner:
                                                Expression:
                                                  expr:
                                                    kind:
                                                      Return:
                                                        value:
                                                          kind:
                                                            Int:
                                                              value: 1
                                                          span:
                                                            start: 60
                                                            end: 61
                                                    span:
                                                      start: 53
                                                      end: 61
                                              span:
                                                start: 53
                                                end: 61
                                              separator: ~
                                          close:
                                            start: 62
                                            end: 63
                                      span:
                                        start: 51
                                        end: 63
                                    else_block: ~
                                span:
                                  start: 41
                                  end: 63
                          span:
                            start: 41
                            end: 63
                          separator:
                            start: 63
                            end: 64
                        - inner:
                            Expression:
                              expr:
                                kind:
                                  Binary:
                                    left:
                                      kind:
                                        Identifier:
                                          symbol: 2
                                      span:
                                        start: 69
                                        end: 70
                                    operator:
                                      kind: Mul
                                      span:
                                        start: 71
                                        end: 72
                                    right:
                                      kind:
                                        Call:
                                          callee:
                                            kind:
                                              Identifier:
                                                symbol: 1
                                            span:
                                              start: 73
                                              end: 82
                                          arguments:
                                            open:
                                              start: 82
                                              end: 83
                                            inner:
                                              - inner:
                                                  kind:
                                                    Binary:
                                                      left:
                                                        kind:
                                                          Identifier:
                                                            symbol: 2
                                                        span:
                                                          start: 83
                                                          end: 84
                                                      operator:
                                                        kind: Sub
                                                        span:
                                                          start: 85
                                                          end: 86
                                                      right:
                                                        kind:
                                                          Int:
                                                            value: 1
                                                        span:
                                                          start: 87
                                                          end: 88
                                                  span:
                                                    start: 83
                                                    end: 88
                                                span:
                                                  start: 83
                                                  end: 88
                                                separator: ~
                                            close:
                                              start: 88
                                              end: 89
                                      span:
                                        start: 73
                                        end: 89
                                span:
                                  start: 69
                                  end: 89
                          span:
                            start: 69
                            end: 89
                          separator: ~
                      close:
                        start: 90
                        end: 91
                  span:
                    start: 35
                    end: 91
            span:
              start: 7
              end: 91
      span:
        start: 0
        end: 91
diagnostics: []
